/*
 * Copyright (c) 2020-2020, Exactpro (Exactpro Systems Limited)
 * www.exactpro.com
 * Build Software to Test Software
 *
 * All rights reserved.
 * This is unpublished, licensed software, confidential and proprietary  
 * information which is the property of Exactpro Systems Limited or its licensors.
 */

syntax = "proto3";
package th2;

import "infra.proto";
import "common.proto";
import "rhbatch.proto";
import "google/protobuf/timestamp.proto";

option java_multiple_files = true;
option java_package = "com.exactpro.th2.act.grpc";

service HandWinAct {
    rpc register(RhTargetServer) returns (ActResponse) {
    }
    rpc unregister(RhSessionID) returns (ActResponse) {
    }

    rpc openApplication (ApplicationDetails) returns (ActResponse) {
    }

    rpc login (LoginDetails) returns (ActResponse) {
    }

    rpc createNewCareOrder (NewCareOrderDetails) returns (ActResponse) {
    }

    rpc acceptOrder (NewCareOrderDetails) returns (ActResponse) {
    }

    rpc getOrderFields (GetOrderFieldsRequest) returns (ActResponse) {
    }

    rpc getChildOrderStatus (CheckChildOrderRequest) returns (ActResponse) {
    }

    rpc directOrder (DirectOrderDetails) returns (ActResponse) {
    }

    rpc closeApplication (BaseRequest) returns (ActResponse) {
    }

    rpc verifyEntities (VerificationDetails) returns (ActResponse) {
    }

    rpc createRFQTileOrder (RFQTileOrderDetails) returns (ActResponse) {
    }

    rpc getDefaultFXValue(DefaultFXValueDetails) returns (ActResponse) {
    }

    rpc getOrderAnalysisAlgoParameters(GetOrderAnalysisAlgoParametersRequest) returns (ActResponse) {
    }
}

message GetOrderFieldsRequest {
    string id = 1;
    RhSessionID sessionID = 2;
    EventID parentEventId = 3;
    repeated ExtractionFieldsDetails details = 4;
}

message CheckChildOrderRequest {
    string id = 1;
    DirectOrderDetails directOrderDetails = 2;
    repeated ExtractionFieldsDetails child = 3;
}

message GetOrderAnalysisAlgoParametersRequest {
    string id = 1;
    RhSessionID sessionID = 2;
    EventID parentEventId = 3;
    repeated ParametersExtractionDetails details = 4;
}

message VerificationDetails {
    message VerificationFields {
        string printedName = 1;
        string actualPath = 2;
        string expectedValue = 3;
    }

    RhSessionID sessionID = 1;
    EventID parentEventId = 2;
    string extractionId = 3;
    string verificationName = 4;
    repeated VerificationFields fields = 5;
}

message ExtractionFieldsDetails {
    string name = 1;
    string colName = 2;
}

message ParametersExtractionDetails {
    string name = 1;
    string paramName = 2;
}

message ApplicationDetails {
    RhSessionID sessionID = 1;
    EventID parentEventId = 2;
    string workDir = 3;
    string applicationFile = 4;
}

message LoginDetails {
    RhSessionID sessionID = 1;
    EventID parentEventId = 2;
    string username = 3;
    string password = 4;
    string mainWindowName = 5;
    string loginWindowName = 6;
}

message NewCareOrderDetails {
    RhSessionID sessionID = 1;
    EventID parentEventId = 2;
    string instrLookupSymbol = 3;
    string limitPrice = 4;
    string quantity= 5;
    string priceType = 6;
    string client = 7;
    string desk = 8;
}

message DirectOrderDetails {
    RhSessionID sessionID = 1;
    EventID parentEventId = 2;
    NewCareOrderDetails orderDetails = 3;
    string qtyPercentage = 4;
}

message RFQTileOrderDetails {

    enum Action {
        BUY = 0;
        SELL = 1;
    }

    RhSessionID sessionID = 1;
    EventID parentEventId = 2;
    string firstCurrency = 3;
    string secondCurrency = 4;
    Action action = 5;
    bool changeCurrency = 6;
    google.protobuf.Timestamp nearDate = 7;
    google.protobuf.Timestamp farDate = 8;
    string venue = 9;
    string nearTenor = 10;
}

message DefaultFXValueDetails {

    message VerificationDetails {
        string name = 1;
        bool isSingleElement = 2;
    }

    string id = 1;
    RhSessionID sessionID = 2;
    EventID parentEventId = 3;
    string extractedField = 4;
    VerificationDetails verificationDetails = 5;
}
